### PR TITLE
Introduce event protocol to avoid otherwise inevitable suffering

### LINKED ISSUE #1233: Implement a testing ensemble
We need ensembles (currently only prefect and legacy) to be a bit more sophisticated than merely implementing `evaluate` and `cancel`. They need to communicate to the evaluator as well.

We have a base `Ensemble` that "implements" the methods, but does not communicate. This makes some of our code awkward to ensure that tests pass.

Should a `TestEnsemble` be implemented that:
- appears to the `EnsembleEvaluator` to be a proper ensemble
- can be configured to 1) fail nominally, 2) fail unexpectedly (crash, network errors, split brain e.g.) 3) be extremely sluggish

❓

### LINKED ISSUE #1235: Define user-evaluator-dispatch protocol
We have a bunch of events flying back and forth. Protocol is the way to make sure the code involved does not descend into chaos. Chaos breeds fear, fear leads to anger anger leads to hate hate leads to suffering.

**Background**
As we develop ERT for the cloud, communication between the different actors: _user_, _Ensemble Evaluator_ (evaluator) and _dispatchers_ will evolve. Extending or changing the current communication is not easy, since reasoning about the current communication requires either reading the implementation, reading the testing code, or run actual experiments. _There is an implicit protocol in the communication between user-evaluator-dispatch that should be made explicit._

**What do we want to define**
- dispatchers-evaluator communication: dispatchers are the steps that are being executed on the cluster. They mainly communicate status and metrics. Currently the communication is one-way, but two-way communication is probably desirable (e.g. you've been running for too long, restart now).
- evaluator-user communication: The evaluator user can be ERT, ert3 or some other external service. The evaluator communicates status, metrics and results. The users tells the evaluator what it needs (cancel, results, etc). This communication is currently two-way, since the user can instruct the evaluator (stop, cancel).
- what concrete events (commands and queries) are sent, and what concrete events are the responses.

**Suggested approach**
Using _[Consumer-Driven Contracts](https://martinfowler.com/articles/consumerDrivenContracts.html)_ to define the user-evaluator-dispatch communication protocol, can provide us with both a way to reason about how the actors interact, as well as programmatic testing of whether or not actors communicate as expected. These contracts are inferred from consumption of "key business" functionality. In other words, unit testing of a _consumer_ could yield a _consumer contract_ based on its use of a _provider_. The sum of all consumer contracts for a given provider becomes the the lower bounds of the _provider contract_.

**Implementation**
Surveying the Consumer-Driven Contracts community, it's heavily geared towards HTTP APIs. E.g. the [Pact spec v3](https://github.com/pact-foundation/pact-specification/tree/version-3) has something called `messages` but it seems like an after thought, as well as insisting on one-way communication. Other libraries (Pact-JVM) also insist on one-way communication.

We could say NIH and make a small code [sic] that allows us to 

- easily define pacts using [pactman](https://github.com/reecetech/pactman) like syntax. (Note: pactman does not implement the message part of the Pact spec v3 https://github.com/reecetech/pactman/issues/68).
- run these as normal pytest
- generate some pact file
- replay said pact file generated from testing a consumer, on a provider
- use pact file to generate documentation
- use pact file(s) to derive provider contracts
- Remove several individual `mock_ws` implementations as they can be replaced by creating a `Pact`.
  - https://github.com/equinor/ert/blob/master/tests/ensemble_evaluator/test_async_queue_execution.py#L10
  - https://github.com/equinor/ert/blob/master/tests/ensemble_evaluator/test_nfs_adoptor_integration.py#L30
  - https://github.com/equinor/ert/blob/4b2dd562d1de33672ff5207b5b4cb8a200b1509e/tests/ensemble_evaluator/conftest.py#L145

### PR DESCRIPTION
**Pull request by jondequinor**
**Issue**
Resolves #1235
Resolves #1233

**Approach**
See #1235, which describes the approach

### CODE REVIEW

File: tests/ensemble_evaluator/test_monitor.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+from tests.narrative_proxy import NarrativeProxy
+from tests.narratives import (
+    monitor_failing_ensemble,
+    monitor_failing_evaluation,
+    monitor_successful_ensemble,
+)
+
+
+@pytest.mark.consumer_driven_contract_test
+def test_verify_narratives2(make_ee_config):
```
Comment (sondreso) [inline tests/ensemble_evaluator/test_monitor.py:25]: Why the 2?

File: tests/ensemble_evaluator/test_monitor.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    ensemble.join()
+
+
+def test_monitor_stop(evaluator):
+    snapshot = None
+    with evaluator.run() as monitor:
+        for event in monitor.track():
+            snapshot = Snapshot(event.data)
+            break
+    assert snapshot.get_status() == ENSEMBLE_STATE_STARTED
```
Comment (sondreso) [inline tests/ensemble_evaluator/test_monitor.py:115]: Is the name of this test correct? Seems like we are testing that the ensemble have started and not stopped 🤔

File: tests/narrative.py [hunk 3]
```diff
@@ -0,0 +1,740 @@
...
+            proto=proto,
+            hostname=hostname,
+            port=port,
+            path=path,
+            base_uri=base_uri,
+        )
+
+
+class ReMatch:
+    # TODO: make regex a re.Pattern after migrating away from 3.6.
```
Comment (sondreso) [inline tests/narrative.py:111]: Open issue with label `EOL Python 3.6` or something similar?

File: tests/narrative.py [hunk 4]
```diff
@@ -0,0 +1,740 @@
...
+        self.direction = None
+        self.ce_serializer = ce_serializer
+
+    def json(self) -> dict:
+        ret = {
+            "provider_states": [state for state in self.provider_states],
+            "type": self.__class__.__name__[1:].lower(),
+            "events": [event.json() for event in self.events],
+        }
+        if hasattr(self, "terminator"):
```
Comment (sondreso) [inline tests/narrative.py:273]: 🤖

File: tests/narrative.py [hunk 5]
```diff
@@ -0,0 +1,740 @@
...
+    async def verify(
+        self, msg_producer: Callable[[], Tuple[InteractionDirection, str]]
+    ):
+        for event in self.events:
+            source, msg = await msg_producer()
+            assert source.represents(
+                self
+            ), f"Wrong direction {source} when expecting {self}\n Got: {msg} instead"
+            self.assert_matches(event, self.ce_serializer.from_json(msg))
+        print("OK", self.scenario)
```
Comment (sondreso) [inline tests/narrative.py:295]: Remove

File: docs/rst/manual/architecture/communication_protocol.rst [hunk 6]
```diff
@@ -0,0 +1,4 @@
+Communication Protocol
+======================
+
+.. ert_narratives::
```
Comment (sondreso) [inline docs/rst/manual/architecture/communication_protocol.rst:4]: 🌟 🥇 🌟

**Review by sondreso — COMMENTED** [review]
Look very good! Was not sure if it was completely ready for review, but added some comments 😄

File: tests/ensemble_evaluator/test_monitor.py (hunk 1 above)
Comment (jondequinor) [reply tests/ensemble_evaluator/test_monitor.py:25]: Relic! Will be removed

File: tests/ensemble_evaluator/test_monitor.py (hunk 2 above)
Comment (jondequinor) [reply tests/ensemble_evaluator/test_monitor.py:115]: Pretty redundant. Will remove.

**Comment by jondequinor** [comment]
@frode-aarstad @mortalisk  anything you'd like to do with this branch before merging? Due to inactivity I've assumed that you were done.

**Review by sondreso — APPROVED** [review]
:ship: :1st_place_medal:

**Pull request merged**